File: 02_agent_patterns/01_agent_loop_basics/agent-loop-target/go-example/async_counter.go
package agentloop

import (
	"runtime"
	"sync"
)

// CountEvents processes n events concurrently and returns how many were counted.
//
// Each event runs in its own goroutine and bumps a shared counter. After all
// goroutines finish, the counter must equal n — no updates may be lost to
// interleaving, and there must be no data race (verify with `go test -race`).
func CountEvents(n int) int {
	counter := 0
	var wg sync.WaitGroup
	for i := 0; i < n; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			c := counter
			runtime.Gosched()
			counter = c + 1
		}()
	}
	wg.Wait()
	return counter
}

File: 02_agent_patterns/01_agent_loop_basics/agent-loop-target/agent_loop_target/async_counter.py
"""Bug 5 (race condition in async code): count events processed concurrently."""

import asyncio


async def count_events(num_events):
    """Process ``num_events`` events concurrently and return how many were counted.

    Each event handler runs as a coroutine and bumps a shared counter. After all
    handlers finish, the counter must equal ``num_events`` — no updates may be
    lost to interleaving.

    Args:
        num_events: the number of events to process.

    Returns:
        The final counter value, which should equal ``num_events``.
    """
    counter = 0

    async def handle_one():
        nonlocal counter
        current = counter
        await asyncio.sleep(0)
        counter = current + 1

    await asyncio.gather(*(handle_one() for _ in range(num_events)))
    return counter
